dataflow/src/org/checkerframework/dataflow/analysis/Analysis.java

```diff
@@ -1,752 +1,61 @@
-     * flow graph. The transfer function is set later using
-     * {@code setTransferFunction}.
-     */
-    public Analysis(ProcessingEnvironment env) {
-        this.env = env;
-        types = env.getTypeUtils();
-    }
-
-    /**
-     * Construct an object that can perform a org.checkerframework.dataflow analysis over a control
-     * flow graph, given a transfer function.
-     */
-    public Analysis(ProcessingEnvironment env, T transfer) {
-        this(env);
-        this.transferFunction = transfer;
-    }
-
-    public void setTransferFunction(T transfer) {
-        this.transferFunction = transfer;
-    }
-
-    public T getTransferFunction() {
-        return transferFunction;
-    }
-
-    public Types getTypes() {
-        return types;
-    }
+public interface Analysis<
+        V extends AbstractValue<V>, S extends Store<S>, T extends TransferFunction<V, S>> {
 
-    public ProcessingEnvironment getEnv() {
-        return env;
+    public static enum Direction {
+        FORWARD,
+        BACKWARD
     }
 
-    /**
-     * Perform the actual analysis. Should only be called once after the object
-     * has been created.
-     */
-    public void performAnalysis(ControlFlowGraph cfg) {
-        assert isRunning == false;
-        isRunning = true;
-
-        init(cfg);
-
-        while (!worklist.isEmpty()) {
-            Block b = worklist.poll();
-
-            switch (b.getType()) {
-                case REGULAR_BLOCK:
-                    {
-                        RegularBlock rb = (RegularBlock) b;
-
-                        // apply transfer function to contents
-                        TransferInput<A, S> inputBefore = getInputBefore(rb);
-                        currentInput = inputBefore.copy();
-                        TransferResult<A, S> transferResult = null;
-                        Node lastNode = null;
-                        boolean addToWorklistAgain = false;
-                        for (Node n : rb.getContents()) {
-                            transferResult = callTransferFunction(n, currentInput);
-                            addToWorklistAgain |= updateNodeValues(n, transferResult);
-                            currentInput = new TransferInput<>(n, this, transferResult);
-                            lastNode = n;
-                        }
-                        // loop will run at least one, making transferResult non-null
-
-                        // propagate store to successors
-                        Block succ = rb.getSuccessor();
-                        assert succ != null
-                                : "regular basic block without non-exceptional successor unexpected";
-                        propagateStoresTo(
-                                succ, lastNode, currentInput, rb.getFlowRule(), addToWorklistAgain);
-                        break;
-                    }
+    public Direction getDirection();
```

Comment:
Now that this is an interface, having good documentation for the interfaces is even more important. Most of these methods already existed, so copy over the documentation for them.